### PR TITLE
Normalize scripts object in manifest with relative bin path

### PR DESCRIPTION
**Pull request by verth**
**Summary**

The scripts object in the package.json is not normalized as with npm. The path `./node_modules/.bin/script` is causing issues on Windows. Strip of the relative path with regex as it's not explicitly needed.

Fixes: 
* https://github.com/yarnpkg/yarn/issues/1657
* https://github.com/yarnpkg/yarn/issues/1606

**Test plan**

* Automated: New tests for normalizing scripts have been added
* Manual:

Use a package.json with a relative bin script (can be done on any OS)

```
{
    "name": "",
    "version": "",
    "scripts": {
        "test": "./node_modules/.bin/webpack --watch"
    }
}
```

Notice that the executed command in the output has been normalized:

```
$ yarn run test
yarn run v0.17.0-0
$ webpack --watch
```

### CODE REVIEW

File: src/util/normalize-manifest/util.js [hunk 1]
```diff
@@ -64,6 +64,22 @@ export function normalizePerson(person: mixed): mixed | PersonObject {
...
+  const relativeBinPath = /^(\.[\/\\])?node_modules[\/\\].bin[\\\/]/;
+  if (script.match(relativeBinPath)) {
+    script = script.replace(relativeBinPath, '');
+  }
+  return script;
+}
+
+export function normalizeScripts(scripts: Object): Object {
+  const normalizedScripts = {};
+  Object.keys(scripts).forEach((scriptName) => {
```
Comment (wyze) [inline src/util/normalize-manifest/util.js]: Change this to `for...in` please.

**Comment by wyze** [comment]
Also, could you show another test where it has something like this in the scripts:

```js
{
  "scripts": {
    "build-prod": "NODE_ENV=production ./node_modules/.bin/webpack"
  }
}
```

As I think this is something we would want normalized, correct?

**Comment by wyze** [comment]
Also, I don't think this addresses #1697 as that has to do with the `bin` field in the `package.json` and this just touches `scripts`.

File: src/util/normalize-manifest/util.js (hunk 1 above)
Comment (verth) [inline src/util/normalize-manifest/util.js]: Done

**Comment by verth** [comment]
Thanks for the review! I addressed your remarks.

Concerning: `"NODE_ENV=production ./node_modules/.bin/webpack"`

Nice use-case, as you would probably have expected this does not get normalized. Made me think as-well of other use-cases that will (not fully) be normalized, ex:
* `"./node_modules/.bin/script1 && ./node_modules/.bin/script2"`
* `"./node_modules/.bin/script1 | ./node_modules/.bin/script2"`
* Actually anything were the relative path to normalize is not first

We would indeed want these normalized as-well. I tested npm too, and these more specific use-cases fail with `npm run` too, which of course is no reason to not fix them. 

What would be the best way to continue here in your opinion? Create a follow-up issue for these cases, as this will already address the reported issues and bring more consistency with npm, so it's not blocking, or not?

**Comment by kittens** [comment]
Can you link to the part of the npm source code that does this normalisation? While I believe this fixes the original issue I'm pretty sure it's more pathological than just that.

**Comment by verth** [comment]
[run-script from npm](https://github.com/npm/npm/blob/d081cc6c8d73f2aa698aab36605377c95e916224/lib/run-script.js#L6) uses [read-package-json](https://github.com/npm/read-package-json/blob/master/read-json.js#L128-L144).

**Comment by pumano** [comment]
@verth Thanks for patch!

**Comment by whitecolor** [comment]
It doesn't fix https://github.com/yarnpkg/yarn/issues/1729
It should be possible to run any executable using `yarn run` scripts.

Npm too [can not run such scripts currenlty](https://github.com/npm/npm/issues/13789).

**Comment by bestander** [comment]
I think we fixed that with different path normalization later on.
Feel free to reopen again if it covers a new tests case.
Sorry for forgetting about this PR.

**Pull request closed without merging**